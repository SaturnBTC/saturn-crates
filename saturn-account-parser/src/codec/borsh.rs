use arch_program::account::AccountInfo;
use arch_program::program_error::ProgramError;
use borsh::{BorshDeserialize, BorshSerialize};
use std::io::Cursor;

/// Classical Borsh serialisation codec.
pub struct BorshCodec;

impl BorshCodec {
    /// Deserialises the shard held in `account` using Borsh.
    pub fn load<S>(account: &AccountInfo<'_>) -> Result<S, ProgramError>
    where
        S: BorshDeserialize + BorshSerialize,
    {
        let data = account.try_borrow_data()?;
        S::try_from_slice(&data).map_err(|_| ProgramError::InvalidAccountData)
    }

    /// Serialises `shard` back into `account.data` using Borsh.
    pub fn store<S>(account: &AccountInfo<'_>, shard: &S) -> Result<(), ProgramError>
    where
        S: BorshSerialize,
    {
        let mut data = account.try_borrow_mut_data()?;
        // Write directly into the account's data buffer via an in-place cursor.
        let mut cursor = {
            let slice: &mut [u8] = &mut *data;
            Cursor::new(slice)
        };
        shard
            .serialize(&mut cursor)
            .map_err(|_| ProgramError::InvalidAccountData)?;
        Ok(())
    }
}

/// Anchor-style Borsh account wrapper mirroring Anchor's `Account<'info, T>`.
#[allow(clippy::module_name_repetitions)]
pub struct Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    account: &'a AccountInfo<'a>,
    data: T,
}

impl<'a, T> Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    pub fn load(account: &'a AccountInfo<'a>) -> Result<Self, ProgramError> {
        let data = BorshCodec::load::<T>(account)?;
        Ok(Self { account, data })
    }

    /// Returns a reference to the underlying `AccountInfo` object.
    pub fn info(&self) -> &'a AccountInfo<'a> {
        self.account
    }

    /// Convenience helper that clones the underlying `AccountInfo`.
    pub fn clone_account(&self) -> AccountInfo<'a> {
        self.account.clone()
    }
}

impl<'a, T> Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize + Default,
{
    pub fn init(account: &'a AccountInfo<'a>) -> Result<Self, ProgramError> {
        let default_val: T = Default::default();
        BorshCodec::store::<T>(account, &default_val)?;
        Ok(Self {
            account,
            data: default_val,
        })
    }
}

impl<'a, T> core::ops::Deref for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.data
    }
}

impl<'a, T> core::ops::DerefMut for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.data
    }
}

impl<'a, T> Drop for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn drop(&mut self) {
        let _ = BorshCodec::store::<T>(self.account, &self.data);
    }
}

/// Alias so callers can write `BorshAccount<'info, T>` like in Anchor.
pub type BorshAccount<'a, T> = Account<'a, T>;

// Allow treating `BorshAccount` as an `AccountInfo` via `AsRef`.
impl<'a, T> AsRef<arch_program::account::AccountInfo<'a>> for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn as_ref(&self) -> &arch_program::account::AccountInfo<'a> {
        self.account
    }
}
