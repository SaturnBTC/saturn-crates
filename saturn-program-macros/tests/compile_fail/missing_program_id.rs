use borsh::{BorshDeserialize, BorshSerialize};
use saturn_account_macros::Accounts;
use saturn_account_parser::Account;
use saturn_account_parser::Context;
use saturn_program_macros::saturn_program;

#[derive(Accounts)]
pub struct DummyAcc<'info> {
    #[account(signer, mut)]
    caller: Account<'info, u64>,
}

#[derive(BorshSerialize, BorshDeserialize)]
pub struct Instr {
    pub value: u64,
}

#[saturn_program]
mod my_prog {
    use super::*;
    // Intentionally **no** declare_id! or ID constant here.

    pub fn dummy<'info>(
        _ctx: Context<'info, DummyAcc<'info>>,
        _ix: Instr,
    ) -> Result<(), arch_program::program_error::ProgramError> {
        Ok(())
    }
}

fn main() {}
