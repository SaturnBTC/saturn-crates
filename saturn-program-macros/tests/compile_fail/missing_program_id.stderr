error[E0425]: cannot find value `ID` in the crate root
  --> tests/compile_fail/missing_program_id.rs:18:1
   |
18 | #[saturn_program]
   | ^^^^^^^^^^^^^^^^^ not found in the crate root
   |
   = note: this error originates in the attribute macro `saturn_program` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused import: `saturn_account_parser::Context`
 --> tests/compile_fail/missing_program_id.rs:4:5
  |
4 | use saturn_account_parser::Context;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
